Components: Add a PageHeader / ActionBar component

ItsJonQ opened the pull request:
## Still work in progress! Creating this PR early as a draft.

👋 Hi there! I'm my journey to help improve `@wordpress/components`, I've started working on adding common components (and primitives) that I feel like are missing in the current library. Common components that one would typically find in other libraries/frameworks, like Bootstrap, Material UI, etc...

One of them is some sort of top/action/page header. Something responsible for indicating context, and containing a handful of actions (links, buttons, maybe tab bars).

### Deploy Preview
https://deploy-preview-18141--gutenberg-playground.netlify.com/design-system/components/?path=/story/pageheader--default

☝️ Check out the Storybook preview to quickly check out the changes 😄 

### Designs

@davewhitley has been kind enough help me with some initial design ideas.

#### Simple
![Screen Shot 2019-10-28 at 12 52 51 PM](https://user-images.githubusercontent.com/2322354/67703234-7fb64c00-f989-11e9-838b-92e622b3e0de.png)

In the above example, there's a title, with some actions on the left and right side.

#### Complex
![Screen Shot 2019-10-28 at 12 52 37 PM](https://user-images.githubusercontent.com/2322354/67703248-83e26980-f989-11e9-9dac-547894af5c05.png)

In the above more complex example, there's a subtitle, as well as an area to add a tab bar-like UI.

### Examples

From looking at different libraries, it looks like there are a couple of use cases:

* A dedicated top nav bar (e.g. [Bootstrap](https://getbootstrap.com/docs/4.0/components/navbar/))
* A dedicated page header bar (e.g. [Shopify Polaris](https://polaris.shopify.com/components/structure/page) . [Salesforce Lightning](https://lightningdesignsystem.com/components/page-headers/))
* A bar that (kinda) combines both (e.g. [Material](https://material.io/components/app-bars-top/#anatomy))

### What this is not

I don't think this component should be used as the main top navbar - something with nav links. Example, [Bootstrap's nav bar](https://getbootstrap.com/docs/4.0/components/navbar/).

### Where this can be used

It may be used to organize the actions at the top of Gutenberg's header bar area.
Also, it would be a handy component for folks to use in custom Gutenberg block creation, or for general purpose use outside of Gutenberg.

### Notes

<details><summary>Title/Subtitle Font Styles</summary>
<p>

At the moment, I haven't added font styles for `title` and `subtitle`. Yet. I didn't know how hands-off we wanted to be with these elements.

</p>
</details>

### Thoughts ❤️

This is just an idea! A proposal :). I'm very open to thoughts re: purpose, naming, etc...

Overall, I think we need **something** to fill in this action bar/page header UI gap.

Would love to hear what everyone thinks. For now, I've just created some bare bones components + styles to get the draft going.

Commit:
Add faker to generate mock data for storybook stories

Commit:
WIP: Create initial PageHeader component

ItsJonQ commented:
Deploy preview!

https://5db72b6369b618000bd76833--gutenberg-playground.netlify.com/design-system/components/?path=/story/pageheader--default

Commit:
WIP: Adding stories for PageHeader. Testing styles for alignment and position

ItsJonQ commented:
Adding Story examples:

![Screen Capture on 2019-10-28 at 16-13-51](https://user-images.githubusercontent.com/2322354/67714407-391f1c80-f99e-11e9-8b36-cfd45bf9509f.gif)

My plan is to create all the various use cases for this component. This has already presented some UI and component API questions.

simison commented on packages/components/src/page-header/stories/index.js:
```diff
@@ -0,0 +1,98 @@
+
+export const _default = () => {
+	const title = text( 'title', 'Title' );
+	const subtitle = text( 'subtitle', faker.lorem.paragraph() );
```
Feels a bit much to introduce `faker` just for random lorem ipsum; I think it would be fine if this is always the same-ipsum instead of random-ipsum  :-)

ItsJonQ replied:
@simison Sounds good! Will remove 🤗

Commit:
Remove Faker. Update story to show right vs bottom placement

ItsJonQ commented:
cc'ing @diegohaz. I remember you were looking into `Toolbar` stuff. This is kinda creeping into that territory 😂

youknowriad commented:
Thanks for the PR 👍 

Do you think this is something that can be used for the Gutenberg header even if it's capable of handling more use-cases? As always, I feel more confident adding components when they are actually used in the plugin (it's not a hard restriction though).

ItsJonQ commented:
@youknowriad My pleasure!

> Do you think this is something that can be used for the Gutenberg header even if it's capable of handling more use-cases

![Screen Shot 2019-10-29 at 09 21 56](https://user-images.githubusercontent.com/2322354/67770889-a29a3c00-fa2d-11e9-9a44-4d570072df45.png)

Just to clarify! Is this the Gutenberg header? If so, then this component can be used there :).

`<tangent>`

The thing I'm unsure of is.. is how to distinguish a Toolbar/ActionBar from something like a PageHeader 🤔 . 

Do PageHeaders always need a title? If it doesn't have a title... does that make it a Toolbar/ActionBar?

Still something I'm playing with. cc'ing @diegohaz + @davewhitley 

`</tangent>`

> As always, I feel more confident adding components when they are actually used in the plugin 

Of course! That makes sense :). In this case, it feels like the `PageHeader` can be used in the main Gutenberg Header (depending on the semantics of Toolbar vs PageHeader).

If not...

> (it's not a hard restriction though)

I'm glad you mentioned that (❤️ ). I think it's valuable to add components that folks would expect from a library/component set. From a Gutenberg context, it would enable folks to be more creative with their custom blocks (and us too!), because they have more UI bits to play with.

ItsJonQ commented:
Update!

## Same, but different

In working through these components, I think I'm noticing 2 use cases:

* One with a title/subtitle
* One without

Even though they're almost identical in what they do (render stuff, in a white box, with somethings on the left and right), the **understanding of their purpose within the context of UI is different**.

From a (React component) code perspective...

I think we need to create something smaller. e.g. `Toolbar` or `ActionBar`.
Something more generally that can be refined/customized to construct something like `PageHeader`.

We took this approach with `Tooltip`!

`Tooltip` and `Popover` are basically identical in function. But there are subtle difference that distinguish where you'd use them. Therefore, you have a component for both.

But, under the hood, `Tooltip` is powered by `Popover`! 

I think the same approach can work here as well 👍

MichaelArestad commented:
I think, from a programmatic standpoint, both the Gutenberg Toolbar and the PageHeader component have much in common. The question is whether they should or should not be the same component.

I believe a component should be focused in purpose and simple to implement.

In the context of these two design patterns, their purposes and needs differ enough to warrant each being their own separate component. Perhaps they could be built behind the scenes sharing some code, but to someone implementing either in their design, they should be separate.

There are secondary benefits for them being separate components. As separate components, they will be more flexible to iterate as there are fewer possible prop-based variants in each. This reduces the complexity of designing and developing them down the road.

The Toolbar can be optimized to be a toolbar without concern for PageHeader needs. The PageHeader can be optimized to be a page header without concern for Toolbar needs.

ItsJonQ commented:
@MichaelArestad 😍 Thank you for sharing your insight! It makes me feel like I'm heading in the right direction, haha.

Commit:
Create ActionBar component to construct PageHeader

Commit:
Adjust arrow icon in ActionBar story.

Adjust styles to add gap

ItsJonQ commented:
Update! with the latest 2 updates, I've created another component (`ActionBar`). It's a general UI component designed to render items (typically actions) and content in a white box with a bottom border.

I then used `ActionBar` to compose `PageHeader`.

This is feeling better! 

While doing this, I'm noticing another pattern starting to emerge.
Will push up that update when I can :)

Commit:
Remove unused prop in ActionBarItem

Commit:
Export ActionBar components from main components/index.js

jameskoster commented:
> I've created another component (ActionBar)

I think this makes a lot of sense. Especially as the ActionBar could have applications outside of the page header, for example the block toolbar:

![toolbar](https://cldup.com/uPFHmYvVP0-3000x3000.png)

Commit:
Add generic Flex component. Refactor ActionBar and PageHeader

MichaelArestad commented:
The Flex component is well done and something that would be great in our other applications. Nice work!

Commit:
ActionBar: Add sizes + Gutenberg Bar example

ItsJonQ commented:
Update:

## Gutenberg Top Bar example

Putting `ActionBar` to the test! I was able to recreate the top Gutenberg bar relatively easily with this new component 😍 

<img width="719" alt="Screen Shot 2019-10-30 at 12 43 28 PM" src="https://user-images.githubusercontent.com/2322354/67879532-87096100-fb13-11e9-82b7-a6035b57cf22.png">

The markup looks like this:

(Note: `Tooltip` removed to keep example short)

```jsx
<ActionBar className="action-bar-story-example" size="small">
  <ActionBarItem>
    <Flex justify="left" gap="extraSmall">
      <IconButton icon="insert" />
      <IconButton icon="undo" />
      <IconButton icon="redo" />
      <IconButton icon="info" />
      <IconButton icon="menu" />
    </Flex>
  </ActionBarItem>
  <ActionBarItem>
    <Flex justify="left" gap="extraSmall">
      <IconButton icon="admin-generic" />
      <IconButton icon="ellipsis" />
    </Flex>
  </ActionBarItem>
</ActionBar>
```

It's not 100% complete, as the Gutenberg bar has responsive styles. But, I think it's a great demonstration for how UI like the Gutenberg bar can be constructed using `ActionBar` and `Flex`

jameskoster commented:
The Gutenberg bar still has the back action on the left, but that's looking awesome @ItsJonQ !

ItsJonQ commented:
@jameskoster Thank you!

That screenshot raises an interesting point...

How do we handle a divider between groups of stuff? 😅 

The a couple of ways I can imagine going about this would be:

* Creating a `<ActionBarDivider />`, allowing you to add wherever you want
* Adding a prop to `<ActionBarItem />` or `<ActionBarBlock />` that'll render a divider (border)
* Create a `<ActionBarGroup />` that'll render a divider when 2 groups are next to each other?

For simplicity, I'm in favour of creating a `<ActionBarDivider />`. It allows for more flexibility when creating UI (from my experience).

It's the same approach I took with [the Card component](https://github.com/WordPress/gutenberg/pull/17963).

Would love thoughts from others though! 🎉

MichaelArestad commented:
> Creating a <ActionBarDivider />, allowing you to add wherever you want
> Adding a prop to <ActionBarItem /> or <ActionBarBlock /> that'll render a divider (border)

For implementation simplicity, I like the prop, but the `ActionBarDivider` is probably more flexible and simpler in terms of styles/etc.

ItsJonQ commented:
@MichaelArestad Thanks for thoughts!!

As an experiment, I'm gonna see how I can create both of these UIs (below), with a Component API that feels good :)

![Screen Shot 2019-10-30 at 16 04 03](https://user-images.githubusercontent.com/2322354/67894513-107a5c80-fb2f-11e9-972c-002d690979e3.png)

Commit:
Refactor ActionBar with divider + support to (maybe) render Toolbar

This update refactors how the styles are defined within ActionBar and
it's sub-components. This allows for greater flexibility, and can be more
easily combined with something like a `Divider`.

An experimental `Toolbar` component was added in the Story to see test
how compatible the `ActionBar` API with a Toolbar like interface.

ItsJonQ commented:
@jameskoster + @MichaelArestad 

Check it out! It can now render both interfaces:

<img width="324" alt="Screen Shot 2019-10-30 at 5 04 22 PM" src="https://user-images.githubusercontent.com/2322354/67899245-45d77800-fb38-11e9-95c0-da0a203f319e.png">

<img width="380" alt="Screen Shot 2019-10-30 at 5 04 28 PM" src="https://user-images.githubusercontent.com/2322354/67899260-4cfe8600-fb38-11e9-8c12-45278743ee75.png">

For the (experimental) Toolbar like interface, I created an (experimental) `Toolbar` component. I just wanted to see how compatible the APIs were. It's just a wrapper for `ActionBar`, with some props configured.

```jsx
/**
 * EXPERIMENT:
 * Creating a Toolbar component using ActionBar
 */
const Toolbar = ( props ) => {
	return (
		<ActionBar
			{ ...props }
			className="action-bar-story-example toolbar"
			dividerSpacing="none"
			innerPadding="extraSmall"
			justify="left"
		/>
	);
};
const ToolbarItem = ActionBarItem;
```

With that, I was able to use it to render the UI in the screenshot/story:

```jsx
<Toolbar>
  <ToolbarItem>
    <Flex justify="left" gap="extraSmall">
      <IconButton icon="insert" />
    </Flex>
  </ToolbarItem>
  <ToolbarItem>
    <Flex justify="left" gap="extraSmall">
      <IconButton icon="undo" />
      <IconButton icon="redo" />
    </Flex>
  </ToolbarItem>
  <ToolbarItem>
    <Flex justify="left" gap="extraSmall">
      <IconButton icon="info" />
      <IconButton icon="menu" />
    </Flex>
  </ToolbarItem>
</Toolbar>
```

Kinda neat!

Note: I'm not married to any of these component names or conventions! I really like the direction this PR is heading. What started off as something for PageHeader transformed into something greater :)

Commit:
Remove unnecessary Flex in Toolbar story

MichaelArestad commented:
> innerPadding="extraSmall"

Nice. Really nice.

The added `Flex` component inside feels less streamlined, but I can see how in this component, extra flexibility (no pun intended) and clarity are desirable over just making the flex work. Could the `Flex` `gap` prop perhaps default to a size based on the `ActionBar` `innerPadding` prop value? You might also make the `justify` prop default to `left`.

afercia commented:
> Do PageHeaders always need a title? 

I can answer this one 🙂 
Any page needs a visible main H1 heading.

In the WordPress admin there's been a lot of work in the last releases to rebuild a decent headings hierarchy that went lost over the years. Today, most of the pages have a main H1 and a good headings hierarchy after the main heading.

In Gutenberg, there was a lengthy discussion about a visible main H1. Due to lack of consensus, all that the accessibility team was able to get was a _visually hidden_ H1. Which is still not ideal, in the accessibility team's opinion. Headings should be visible.

The main heading answer a fundamental question for users: Where am I? What can I do here?
Visible headings benefit all users. Also, all the Ui sections must be identified by headings.

References:

**Headings**
Headings communicate the organization of the content on the page. Web browsers, plug-ins, and assistive technologies can use them to provide in-page navigation.
https://www.w3.org/WAI/tutorials/page-structure/headings/

**Page Regions**
Mark up different regions of web pages and applications, so that they can be identified by web browsers and assistive technologies.
https://www.w3.org/WAI/tutorials/page-structure/regions/

A main H1 is of fundamental importance at the point a `PageHeader` component should make it required.

afercia commented:
Regarding the Gutenberg Top Bar example

> Putting `ActionBar` to the test! I was able to recreate the top Gutenberg bar relatively easily with this new component

Well, that recreates the Gutenberg bar _only visually_. It doesn't rebuild at all the semantics, the ARIA toolbar pattern, arrows navigation etc. I'm a bit surprised explorations like this one completely ignore the necessary semantics and expected interaction, focusing only on the visual part. This is not exactly the kind of _inclusive design_ we aim for in WordPress. I'd kindly invite everyone to take into consideration since the beginning all the necessary web standards and accessibility patterns when designing something new.

These patterns are complets and they're not only visual. They require specific expertise. I do realize there's good in new explorations but at the same time designers should strive to implement standards and accessibility since the beginning. If design doesn't have the necessary expertise, there are specialists in this community who can provide all the necessary knowledge.

afercia commented:
A quick, kind, procedural note.:

I'd like to see any PR in this repository to be accompanied by a related issue. Issues are better exposed to a larger community of contributors and teams. They allow a larger participation and feedback. Instead, proposals that happen only in a PR are not so exposed. I think I already asked for this in the past, but I see  PRs without issues are still happening. 

Personally, I perceive a PR without an accompanying issue as something that was discussed "internally" by a smaller group of contributors. A PR that comes out of the blue and doesn't clarify where the discussion the PR is based on happened. Assuming it happened somewhere. Honestly I don't feel pretty comfortable with this. I think I'm not alone in this feeling and I'd like to invite everyone to consider that larger participation adds great value.

Also, I'd like to remind everyone that any proposal of large changes to any UI should have the "Needs Accessibility Feedback" label. Thank you 🙂

simison commented:
> Check it out! It can now render both interfaces:
> 
> <img width="324" alt="Screen Shot 2019-10-30 at 5 04 22 PM" src="https://user-images.githubusercontent.com/2322354/67899245-45d77800-fb38-11e9-95c0-da0a203f319e.png">
> 
> <img width="380" alt="Screen Shot 2019-10-30 at 5 04 28 PM" src="https://user-images.githubusercontent.com/2322354/67899260-4cfe8600-fb38-11e9-8c12-45278743ee75.png">


Cool! :-) Curious how dropdowns would work technically and visually in both toolbars?

MichaelArestad commented:
> Any page needs a visible main H1 heading.

@afercia That's a good point. That doesn't mean this component needs to have an H1. There are times when it could be used outside of the standard page header context or when the design calls for an H1 elsewhere.

> explorations like this one completely ignore the necessary semantics and expected interaction, focusing only on the visual part.

This is an early build and, as such, will take time to work the necessary semantics and interactions in. We could use help from someone with expertise.

> These patterns are complex and they're not only visual. They require specific expertise. I do realize there's good in new explorations but at the same time designers should strive to implement standards and accessibility since the beginning. If design doesn't have the necessary expertise, there are specialists in this community who can provide all the necessary knowledge.

That's why this is is a group effort. I, for one, welcome your suggestions and contributions.

> I'd like to see any PR in this repository to be accompanied by a related issue. Issues are better exposed to a larger community of contributors and teams. They allow a larger participation and feedback. Instead, proposals that happen only in a PR are not so exposed.

Good point. We can spin one up and will make a point to do so in the future. I know it's frustrating to repeat the basics over and over. So many people in leadership positions (everywhere, not just WordPress) can sympathize. We were all new contributors once and the process took time to learn then -- just imagine what it's like 17 years into the project trying to figure it all out.

ItsJonQ commented:
@MichaelArestad Thank you so much for replying! 💯 to all the things you mentioned.

---

@afercia Thank you for your reply!! @MichaelArestad beat me to replying (traveling today!)

> ...I'm a bit surprised explorations like this one completely ignore the necessary semantics and expected interaction, focusing only on the visual part.... This is not exactly the kind of inclusive design we aim for in WordPress...

Echo'ing a bit of what @MichaelArestad said.

It's still very early for this component (or these sets of components)! I'm still exploring what these pieces are. Accessibility is definitely something we need to built into this lower level/primitive/base components. It's tricky though, as I'm still unsure how these bits will be used yet.

In the most recent update, I created a Toolbar as a test to see if the Components could even do it.
And if so, should it?

I think once we identify that from this, we can create a `Toolbar`, `PageHeader`, or anything else, we will have the markup to apply the a11y support.

> These patterns are complex and they're not only visual. They require specific expertise...

Oh for sure! That's why I'm asking for feedback to see how the Component APIs feel. They why it currently works is similar to other React libraries. And the idea behind those patterns, is similar to ones that Bootstrap and other CSS frameworks use:

e.g. Bootstrap grid:
https://getbootstrap.com/docs/4.0/layout/grid/#how-it-works

Using different pre-made classes (their "components") to create things.

> I'd like to see any PR in this repository to be accompanied by a related issue. Issues are better exposed to a larger community of contributors and teams. They allow a larger participation and feedback. Instead, proposals that happen only in a PR are not so exposed.

My apologies! I wasn't sure of the process. I thought a draft PR might have been appropriate, since it was a draft. Also, I'm able to leverage the Netlify builds which allow for people to see the changes and poke at the Storybook more easily.

I'll be more mindful next time and open up an issue first.

---

I'm very glad you joined the conversation! I think it's important various folks (especially experts) collaborate on a design like this - especially at the early stages.

ItsJonQ commented:
> Cool! :-) Curious how dropdowns would work technically and visually in both toolbars?

@simison Me too! I can give it a shot in the next iteration :)

Commit:
Merge branch 'master' into add/page-header-component

Commit:
Update label for stories

Commit:
Merge branch 'master' into add/page-header-component

ItsJonQ commented:
👋 Hallo! Following up. This is still active :).

I've been coordinating with @diegohaz. It looks like his Toolbar enhancement PR is close to landing:
https://github.com/WordPress/gutenberg/pull/17875

Since the component work has some overlap with his. I was going to wait to see how that PR gets merged before continuing

ItsJonQ commented:
Closing this PR. We can revisit it in the future
